Add options to filter lists_mergeby

vbotka opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Implement parameters 'recursive' and 'list_merge'. The same functionality provided by the filter 'combine' for the dictionaries will be available for the merging of the lists of dictionaries.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
 Fixes #4057
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
lists_mergeby

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
```

vbotka commented:
For the record. Six tests failed because of outdated  azure-pipelines

```
 merge_hash() takes exactly 2 arguments (4 given)
```

felixfontein reviewed (commented):
Thanks for your contribution! Unfortunately, there's a problem with Ansible 2.9 compatiblity. What would be possible is to add some special code which checks for Ansible 2.9, and if that is used then error out if these parameters are set to non-default values; and if they equal the default values, somehow make sure that they are not passed on. That will unfortunately make the code more complex though...

Besides that, this needs a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to). Also you should update the filter documentation in https://github.com/ansible-collections/community.general/blob/main/docs/docsite/rst/filter_guide.rst#merging-lists-of-dictionaries.

felixfontein commented:
> For the record. Six tests failed because of outdated azure-pipelines
> 
> ```
>  merge_hash() takes exactly 2 arguments (4 given)
> ```

They are not outdated, they show exactly what the problem with the PR is: it doesn't work with Ansible 2.9 :-)

vbotka commented on plugins/filter/list.py:
```diff
@@ -1,43 +1,88 @@
                 raise AnsibleFilterError('Elements of list arguments for lists_mergeby must be dictionaries. Found {0!r}.'.format(elem))
             if index in elem.keys():
-                d[elem[index]].update(elem)
+                d[elem[index]].update(merge_hash(d[elem[index]], elem, recursive, list_merge))
```
Thank you for the comments! I created a wrapper to deal with it. It's not such a burden and might be useful in the future. What do you think?

I'll proceed with the changelog and doc. In the doc, my intention is to refer to [Combining hashes/dictionaries](https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries) because the function 'merge_hash' from ansible.utils.vars, used by the filter 'combine', is also used by the filter 'lists_mergeby'. Are there any rules for such cross-referencing?

felixfontein replied:
There are no rules for cross-referencing yet.

felixfontein commented on plugins/filter/list.py:
```diff
@@ -1,43 +1,116 @@
 from collections import defaultdict
 from operator import itemgetter
+from packaging.specifiers import SpecifierSet
+from packaging.version import Version
```
Please use `LooseVersion` from `ansible_collections.community.general.plugins.module_utils.version` instead. `packaging` is not necessarily installed everywhere (it has been a requirement since 2.10, but is not one of 2.9).

felixfontein commented on plugins/filter/list.py:
```diff
@@ -1,43 +1,116 @@
-    if not isinstance(l2, Sequence):
-        raise AnsibleFilterError('Second argument for community.general.lists_mergeby must be list. %s is %s' %
-                                 (l2, type(l2)))
+    if Version(ansible_version) not in vspec:
```
```suggestion
    if LooseVersion(ansible_version) < LooseVersion('2.10'):
```

vbotka replied:
I exchanged SpecifierSet for LooseVersion. Thank you!

vbotka commented:
Would it be possible to review the changes in the docs? I added *examples*, used *literalinclude*, and changed the format of the output to *yaml*. Is it OK to proceed this way?

felixfontein commented on docs/docsite/examples/lists_mergeby/example-001.out:
```diff
@@ -0,0 +1,10 @@
+  list3:
```
Files in this directory will not be found during the docs build. The files must be in docs/docsite/rst/.

vbotka replied:
I'll move it inside the 'rst' directory. Thank you.

felixfontein replied:
I'll try to build the docs to see whether it actually works.

vbotka replied:
> I'll try to build the docs to see whether it actually works.

JFYI. I was running sphinx-view while editing the docs.

felixfontein replied:
Unfortunately it does not work:
```
docsite/rst/collections/community/general/docsite/filter_guide.rst:257: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-001.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:262: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-001.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:269: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-002.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:274: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-002.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:289: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-003.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:294: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-003.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:299: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-004.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:304: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-004.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:309: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-005.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:314: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-005.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:319: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-006.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:324: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-006.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:329: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-007.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:334: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-007.out' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:339: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-008.yml' not found or reading it failed
docsite/rst/collections/community/general/docsite/filter_guide.rst:344: WARNING: Include file 'docsite/rst/collections/community/general/docsite/examples/lists_mergeby/example-008.out' not found or reading it failed
```
Antsibull only copies `.rst` files over (https://github.com/ansible-community/antsibull/blob/1eb595309ec0e01310edd59682e808dde3fc8a3a/src/antsibull/extra_docs.py#L44-L60, https://github.com/ansible-community/antsibull/blob/1eb595309ec0e01310edd59682e808dde3fc8a3a/src/antsibull/extra_docs.py#L194-L206). So for now you have to inline the files instead of using `literalinclude`. (I would still keep the files, so that it's easier to reproduce and to eventually replace the inlined copies with the originals.)

vbotka replied:
This is not only awkward but error-prone as well. Would it be possible to try the *.rst files I submitted?

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,34 @@ This produces:
+If you have two or more lists of dictionaries and want to combine them into a list of merged dictionaries, where the dictionaries are merged by an attribute, you can use the ``lists_mergeby`` filter.
 
-.. code-block:: yaml+jinja
+.. note:: The output of the examples in this section use the yaml callback plugin. Quoting: "Ansible output that can be quite a bit easier to read than the default JSON formatting." See ``shell> ansible-doc -t callback yaml``
```
```suggestion
.. note:: The output of the examples in this section use the YAML callback plugin. Quoting: "Ansible output that can be quite a bit easier to read than the default JSON formatting." See :ref:`the documentation for the community.general.yaml callback plugin <ansible_collections.community.general.yaml_callback>`.
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,34 @@ This produces:
-            path: /foo
-          - name: baz
-            path: /bazzz
+In the example below the lists are merged by the attribute *name*
```
```suggestion
In the example below the lists are merged by the attribute *name*:
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,34 @@ This produces:
+In the example below the lists are merged by the attribute *name*
+
+.. literalinclude:: ../examples/lists_mergeby/example-001.yml
+   :language: yaml
```
The file is not pure YAML. Generally we always use `yaml+jinja` for something that can potentially contain Jinja2.
```suggestion
   :language: yaml+jinja
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,34 @@ This produces:
+It is possible to use a list of lists as an input of the filer
+
+.. literalinclude:: ../examples/lists_mergeby/example-002.yml
+   :language: yaml
```
```suggestion
   :language: yaml+jinja
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,34 @@ This produces:
-    }
+.. versionadded:: 2.0.0
+
+It is possible to use a list of lists as an input of the filer
```
```suggestion
It is possible to use a list of lists as an input of the filer:
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,34 @@ This produces:
+.. literalinclude:: ../examples/lists_mergeby/example-002.out
+   :language: yaml
+
+.. versionadded:: 4.4.0
```
This only makes sense if you document the new options. The above does not do that, so it should not have 4.4.0 as `versionadded`.

vbotka replied:
I'll remove it. Thank you.

vbotka commented:
The PR might be completed, I think. Would you like to add or change anything else?

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/example-003.yml:
```diff
@@ -0,0 +1,28 @@
+  set_fact:
+    list3: "{{ [list1, list2]|
+               community.general.lists_mergeby('name',
+                                               recursive=True) }}"
```
```suggestion
                                               recursive=true) }}"
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/example-004.yml:
```diff
@@ -0,0 +1,29 @@
+  set_fact:
+    list3: "{{ [list1, list2]|
+               community.general.lists_mergeby('name',
+                                               recursive=True,
```
```suggestion
                                               recursive=true,
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/example-005.yml:
```diff
@@ -0,0 +1,29 @@
+  set_fact:
+    list3: "{{ [list1, list2]|
+               community.general.lists_mergeby('name',
+                                               recursive=True,
```
```suggestion
                                               recursive=true,
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/example-006.yml:
```diff
@@ -0,0 +1,29 @@
+  set_fact:
+    list3: "{{ [list1, list2]|
+               community.general.lists_mergeby('name',
+                                               recursive=True,
```
```suggestion
                                               recursive=true,
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/example-007.yml:
```diff
@@ -0,0 +1,29 @@
+  set_fact:
+    list3: "{{ [list1, list2]|
+               community.general.lists_mergeby('name',
+                                               recursive=True,
```
```suggestion
                                               recursive=true,
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/example-008.yml:
```diff
@@ -0,0 +1,29 @@
+  set_fact:
+    list3: "{{ [list1, list2]|
+               community.general.lists_mergeby('name',
+                                               recursive=True,
```
```suggestion
                                               recursive=true,
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,102 @@ This produces:
+.. literalinclude:: examples/lists_mergeby/example-002.out
+   :language: yaml
+
+The filter also accepts two optional parameters: ``recursive`` and ``list_merge``. These parameters are only supported when used with ansible-base 2.10 or ansible-core, but not with Ansible 2.9.
```
```suggestion
The filter also accepts two optional parameters: ``recursive`` and ``list_merge``. These parameters are only supported when used with ansible-base 2.10 or ansible-core, but not with Ansible 2.9. This is available since community.general 4.4.0.
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,102 @@ This produces:
+**list_merge**
+    Is a string, its possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``. It modifies the behaviour of community.general.lists_mergeby when the hashes to merge contain arrays/lists.
+
+The examples below set ``recursive=True`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.
```
```suggestion
The examples below set ``recursive=true`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.
```

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,102 @@ This produces:
+
+The examples below set ``recursive=True`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.
+
+Example **list_merge=replace** (default)
```
```suggestion
Example **list_merge=replace** (default):
```
Please also add this to the others.

felixfontein commented on tests/integration/targets/filter_list/tasks/.yamllint:
```diff
@@ -0,0 +1,20 @@
+---
```
Please remove this file.

vbotka replied:
Done.

felixfontein commented on tests/integration/targets/filter_list/tasks/lists_mergeby_default.yml:
```diff
@@ -0,0 +1,166 @@
+  when: debug_test|d(false)|bool
+  tags: t0
+
+- name: 1.Test lists merged by attribute name
```
```suggestion
- name: 1. Test lists merged by attribute name
```
Same for the others.

vbotka commented:
Oops, bad idea. Work in progress ...

felixfontein commented:
This is how the docs currently look: https://ansible.fontein.de/collections/community/general/docsite/filter_guide.html#merging-lists-of-dictionaries

There also exists https://ansible.fontein.de/collections/community/general/docsite/examples/lists_mergeby/examples.html which generates a warning (`docsite/rst/collections/community/general/docsite/examples/lists_mergeby/examples.rst: WARNING: document isn't included in any toctree`) during docs build.

I'll reorganize the files a bit (also to avoid this extra HTML file) in a comment I'll push to this PR before merging.

felixfontein commented on docs/docsite/rst/filter_guide.rst:
```diff
@@ -247,55 +248,437 @@ This produces:
+**list_merge**
+    Is a string, its possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``. It modifies the behaviour of community.general.lists_mergeby when the hashes to merge contain arrays/lists.
+
+The examples below set ``recursive=true`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.
```
```suggestion
The examples below set ``recursive=true`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See :ref:`Combining hashes/dictionaries <combine_filter>` to learn details about these options.
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/examples.yml:
```diff
@@ -0,0 +1,38 @@
+---
+examples:
+  - label: 'Example list_merge=replace (default):'
```
```suggestion
  - label: 'Example ``list_merge=replace`` (default):'
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/examples.yml:
```diff
@@ -0,0 +1,38 @@
+  - label: 'This produces:'
+    file: example-003.out
+    lang: 'yaml'
+  - label: 'Example list_merge=keep'
```
```suggestion
  - label: 'Example ``list_merge=keep``:'
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/examples.yml:
```diff
@@ -0,0 +1,38 @@
+  - label: 'This produces:'
+    file: example-004.out
+    lang: 'yaml'
+  - label: 'Example list_merge=append'
```
```suggestion
  - label: 'Example ``list_merge=append``:'
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/examples.yml:
```diff
@@ -0,0 +1,38 @@
+  - label: 'This produces:'
+    file: example-005.out
+    lang: 'yaml'
+  - label: 'Example list_merge=prepend'
```
```suggestion
  - label: 'Example ``list_merge=prepend``:'
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/examples.yml:
```diff
@@ -0,0 +1,38 @@
+  - label: 'This produces:'
+    file: example-006.out
+    lang: 'yaml'
+  - label: 'Example list_merge=append_rp'
```
```suggestion
  - label: 'Example ``list_merge=append_rp``:'
```

felixfontein commented on docs/docsite/rst/examples/lists_mergeby/examples.yml:
```diff
@@ -0,0 +1,38 @@
+  - label: 'This produces:'
+    file: example-007.out
+    lang: 'yaml'
+  - label: 'Example list_merge=prepend_rp'
```
```suggestion
  - label: 'Example ``list_merge=prepend_rp``:'
```

felixfontein commented on tests/integration/targets/filter_list/runme.sh:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
I think this file and `runme.yml` can be deleted. They don't seem to achieve anything.

vbotka replied:
Don't delete it. It's a playbook used in *runme.sh*

felixfontein replied:
Yeah, but why? Integration tests don't need a runme.sh and a playbook. That's only useful if you need to install something before the tests run, set environment variables for `ansible-playbook`, etc. For tests that do nothing special, using ansible-test's native way is a lot cleaner.

vbotka replied:
OK. I can live without them.

felixfontein replied:
The scripts in the other roles are there because they cannot simply be used as roles, but some setup has to be done (like installing a Python dependency) that cannot be done in a role where the plugin is used (since plugins are loaded on startup, and most of them check their dependencies on load time). All other tests without a runme.sh are already simple roles.

vbotka commented:
> This is how the docs currently look: https://ansible.fontein.de/collections/community/general/docsite/filter_guide.html#merging-lists-of-dictionaries

The docs look good.
 
> There also exists https://ansible.fontein.de/collections/community/general/docsite/examples/lists_mergeby/examples.html which generates a warning (`docsite/rst/collections/community/general/docsite/examples/lists_mergeby/examples.rst: WARNING: document isn't included in any toctree`) during docs build.

Yes. delete it. The file examples.rst is generated from the template examples.rst.j2 by the playbook.yml. It must be copy/paste into the filter_guide.rst.

> I'll reorganize the files a bit (also to avoid this extra HTML file) in a comment I'll push to this PR before merging.

OK. How can I help?

felixfontein commented:
@vbotka thanks a lot for improving this plugin, and the docs!

vbotka commented:
Thank you for the guidance and patience!

Do you think it might be a good idea to split the file *filter_guide.rst* into files by sections? 

i.e. in *filter_guide.rst* replace 
```
.. contents:: Topics 
```

by

```
.. toctree::
   :maxdepth: 2

   filter_guide_path
   filter_guide_abstract_transformation
   ...
   filter_guide_working_with_times
   filter_guide_working_with_unicode
```

Then split *filter_guide_abstract_transformation* the same way? I volunteer. Do you see any obstacles?
